clang/lib/CIR/CodeGen/LoweringPrepare.cpp

```diff
@@ -865,30 +893,379 @@ void LoweringPreparePass::lowerThreeWayCmpOp(CmpThreeWayOp op) {
+  if (!guard) {
+    // Mangle the name for the guard.
+    SmallString<256> guardName;
+    {
+      llvm::raw_svector_ostream out(guardName);
+      cxxABI.getMangleContext().mangleStaticGuardVariable(varDecl.getRawDecl(),
+                                                          out);
+    }
+
+    // Create the guard variable with a zero-initializer.
+    // Just absorb linkage, visibility and dll storage class from the guarded
+    // variable.
+    guard = createGuardGlobalOp(builder, globalOp->getLoc(), guardName, guardTy,
+                                /*isConstant=*/false, /*addrSpace=*/{},
+                                globalOp.getLinkage(),
+                                getGlobalOp->getParentOfType<cir::FuncOp>());
+    guard.setInitialValueAttr(cir::IntAttr::get(guardTy, 0));
+    guard.setDSOLocal(globalOp.isDSOLocal());
+    guard.setVisibility(globalOp.getVisibility());
+    assert(!::cir::MissingFeatures::setDLLStorageClass());
+    // guard.setDLLStorageClass(globalOp.getDLLStorageClass());
+    // If the variable is thread-local, so is its guard variable.
+    assert(!::cir::MissingFeatures::threadLocal());
+    // guard.setThreadLocalMode(globalOp.getThreadLocalMode());
+    guard.setAlignment(guardAlignment.getAsAlign().value());
+
+    // The ABI says: "It is suggested that it be emitted in the same COMDAT
+    // group as the associated data object." In practice, this doesn't work
+    // for non-ELF and non-Wasm object formats, so only do it for ELF and
+    // Wasm.
+    assert(!::cir::MissingFeatures::setComdat());
+
+    setStaticLocalDeclGuardAddress(varDecl, guard);
+  }
+
+  mlir::Value getGuard =
+      cgm.getBuilder().createGetGlobal(guard, /*threadLocal*/ false);
+  clang::CIRGen::Address guardAddr =
+      clang::CIRGen::Address(getGuard, guard.getSymType(), guardAlignment);
+
+  // Test whether the variable has completed initialization.
+  //
+  // Itanium C++ ABI 3.3.2:
+  //   The following is pseudo-code showing how these functions can be used:
+  //     if (obj_guard.first_byte == 0) {
+  //       if ( __cxa_guard_acquire (&obj_guard) ) {
+  //         try {
+  //           ... initialize the object ...;
+  //         } catch (...) {
+  //           __cxa_guard_abort (&obj_guard);
+  //           throw;
+  //         }
+  //         ... queue object destructor with __cxa_atexit() ...;
+  //         __cxa_guard_release (&obj_guard);
+  //       }
+  //     }
+  //
+  // If threadsafe statics are enabled, but we don't have inline atomics, just
+  // call __cxa_guard_acquire unconditionally.  The "inline" check isn't
+  // actually inline, and the user might not expect calls to __atomic
+  // libcalls.
+  unsigned maxInlineWidthInbits = cgm.getTarget().getMaxAtomicInlineWidth();
+  // "init.end" block
+  mlir::Block *endBlock = nullptr;
+  {
+    mlir::OpBuilder::InsertionGuard guard(builder);
+    endBlock = builder.createBlock(builder.getBlock()->getParent());
```

Comment:
As mentioned above, branches to block are no good, you can start without EH support and add missing features.